h3: add event yielding to poll()

LPardue opened the pull request:
When DATAGRAM support was added, poll() was updated to return
Event::Datagram. We made the decision to prefer returning such
events ahead of stream events (Event::Headers, Event::Data).

There can be cases when an application wants to handle streams
ahead of DATAGRAMS. However, just reordering things inside poll()
wouldn't work for applications that prefer to read DATAGRAMs.

This change introduces a basic yield mechanism into poll().
Poll() starts by processing DATAGRAMs. Once a threshold
count is hit, it moves to processing request
streams. Once that threshold is hit, it moves back to DATAGRAMs.

The thresholds are configured using quiche::h3::Config. By
default there are no thresholds applied, this maintains prior
behavior.

Since this changes poll(), I've added several unit tests for both
simple DATAGRAM sending/receiving, and DATAGRAM poll yielding.

LPardue commented:
I need to do a bit more tweaking on the poll() internals.

LPardue commented:
The tweaks actually ended up changing the API too :) But I think what's there now is conceptually easier to follow.

ghedo commented on src/h3/mod.rs:
```diff
@@ -415,6 +417,8 @@ impl Config {
             max_header_list_size: None,
             qpack_max_table_capacity: None,
             qpack_blocked_streams: None,
+            dgram_poll_threshold: None,
```
Instead of using `Option` here, we could just set the default value to `std::usize::MAX` I think, as it seems like it would simplify things as we wouldn't need to deal with `None` anymore later on.

ghedo replied:
Also, type should probably be `u64` as this doesn't really refer to any "memory-related quantities".

LPardue replied:
I was trying to be clever, let me try the KISS option.

ghedo commented on src/h3/mod.rs:
```diff
@@ -1080,22 +1108,45 @@ impl Connection {
+                return Err(Error::Done);
+            }
+
+            self.stream_poll_count += 1;
```
I think this should be inside the loop, and only increased if we actually return a stream event. Otherwise it seems like it would be increased even when there are no stream events and `Done` is returned?

LPardue replied:
Yeah this seems a bit wonky but during testing the problem I hit is that once you're into processing streams, you have to have a stream event in order to get back to datagrams - but that might never happen e,g, a basic MASQUE session.

The current code is intended to say "I'll try to process only stream events for N calls of poll, after that I'll go back to trying datagrams". It's not a count of the number of events, its the poll count.

ghedo replied:
Normally when `poll()` returns `Done` the application will stop calling it until, e.g. new packets arrive or something. So this behavior might end up delaying the delivery of DATAGRAM events for several RTTs I think.

Instead of doing that, when checking if DATAGRAM can be sent, we can also call `conn.streams.has_readable()` to see if there even are any streams that are readable. so e.g. `if !conn.streams.has_readable() || below dgram threshold {` or something like that.

LPardue replied:
I hadn't appreciated the importance of `Done` before. This causes a mild quandry because there is no outright correct order. Ideally, while there is any DATAGRAM or stream data buffered and ready to be processed, we want poll to return an event even if that "bends" the yield thresholds. We really don't want to have call `poll()` multiple times and have it return `Done` in order to get over the threshold to start processing the data we care about. 

I've tried to address this in https://github.com/cloudflare/quiche/pull/704/commits/91eab5f0dc1ac461e20c84a7e0baf48d329d073e. With that change, the intent is that `poll()` will jump into processing whatever type the connection-global threshold prefers. If there is nothing to do for a given type, switch to the other one. It only returns `Done` when there is nothing to do for either type.

ghedo commented on src/h3/mod.rs:
```diff
@@ -1080,22 +1108,45 @@ impl Connection {
 
         // Process HTTP/3 data from readable streams.
+        if let Some(thresh) = self.stream_poll_threshold {
+            if self.stream_poll_count > thresh {
```
Should probably also move this inside the loop, to after `stream_poll_count` is incremented (see other comment), and change the condition to `>=`. Then we can also avoid returning `Done` here.

ghedo commented on include/quiche.h:
```diff
@@ -443,6 +443,12 @@ void quiche_h3_config_set_qpack_max_table_capacity(quiche_h3_config *config, uin
 // Sets the `SETTINGS_QPACK_BLOCKED_STREAMS` setting.
 void quiche_h3_config_set_qpack_blocked_streams(quiche_h3_config *config, uint64_t v);
 
+// Sets the `quiche_h3_conn_poll()` repetition threshold for DATAGRAMs.
```
```suggestion
// Sets the `quiche_h3_conn_poll()` repetition threshold for DATAGRAM events.
```

ghedo commented on include/quiche.h:
```diff
@@ -443,6 +443,12 @@ void quiche_h3_config_set_qpack_max_table_capacity(quiche_h3_config *config, uin
+// Sets the `quiche_h3_conn_poll()` repetition threshold for DATAGRAMs.
+void quiche_h3_config_set_dgram_poll_threshold(quiche_h3_config *config, size_t v);
+
+// Sets the `quiche_h3_conn_poll()` repetition threshold for streams.
```
```suggestion
// Sets the `quiche_h3_conn_poll()` repetition threshold for stream events.
```

ghedo commented on src/h3/mod.rs:
```diff
@@ -438,6 +442,20 @@ impl Config {
         self.qpack_blocked_streams = Some(v);
     }
+
+    /// Sets the `poll()` repetition threshold for DATAGRAMs.
```
```suggestion
    /// Sets the `poll()` repetition threshold for DATAGRAM events.
```

ghedo commented on src/h3/mod.rs:
```diff
@@ -438,6 +442,20 @@ impl Config {
+        self.dgram_poll_threshold = Some(v);
+    }
+
+    /// Sets the `poll()` repetition threshold for streams.
```
```suggestion
    /// Sets the `poll()` repetition threshold for stream events.
```

ghedo reviewed (changes requested):
I'm somewhat confused by the count reset logic, so some of the comments I made might not work, but I think there are some opportunities to simplify things a bit there.

LPardue commented:
@ghedo I've added a commit that simplifies things and avoids returning Error::Done. Is the intent clearer now?

junhochoi commented on src/h3/mod.rs:
```diff
@@ -406,6 +406,8 @@ pub struct Config {
     max_header_list_size: Option<u64>,
     qpack_max_table_capacity: Option<u64>,
     qpack_blocked_streams: Option<u64>,
+    dgram_poll_threshold: u64,
+    stream_poll_threshold: u64,
```
Does it need to be `u64`?

LPardue replied:
That's what @ghedo suggested in https://github.com/cloudflare/quiche/pull/704#discussion_r519893116

ghedo commented on src/h3/mod.rs:
```diff
@@ -476,6 +480,20 @@ impl Config {
         self.qpack_blocked_streams = Some(v);
     }
+
+    /// Sets the `poll()` repetition threshold for DATAGRAM events.
```
We might want to go into a bit more detail on what these config options do, and why one might want to use them, as it's not immediately evident IMO.

ghedo commented on src/h3/mod.rs:
```diff
@@ -1125,42 +1153,73 @@ impl Connection {
+        if self.dgram_event_count < self.dgram_poll_threshold &&
+            conn.dgram_recv_front_len().is_some()
+        {
+            match conn.dgram_recv_peek(&mut d, 8) {
```
Might be a good idea to refactor this into a separate method, so it can be reused later rather than duplicating the code, though I'm not too sure it would actually simplify things, as you'd still need the `match` here anyway.

LPardue replied:
Good idea, see `process_dgrams() which still returns a Result but handles the boring mapping between error types, which makes the code in `poll()` a little tidier.

ghedo commented on src/h3/mod.rs:
```diff
@@ -1125,42 +1153,73 @@ impl Connection {
-                let flow_id = b.get_varint()?;
-                return Ok((flow_id, Event::Datagram));
-            },
+        if self.dgram_event_count < self.dgram_poll_threshold &&
```
```suggestion
        // Process queued DATAGRAMs if the poll threshold allows it.
        if self.dgram_event_count < self.dgram_poll_threshold &&
```

ghedo commented on src/h3/mod.rs:
```diff
@@ -1125,42 +1153,73 @@ impl Connection {
+            };
+        }
 
-        // Process HTTP/3 data from readable streams.
```
This comment should stay IMO.

ghedo commented on src/h3/mod.rs:
```diff
@@ -1125,42 +1153,73 @@ impl Connection {
+        }
 
-            // TODO: check if stream is completed so it can be freed
+        // no event generated at this point, so try dgram again
```
```suggestion
        // No event generated at this point, so try dgram again.
```

ghedo commented on src/h3/mod.rs:
```diff
@@ -1125,42 +1153,73 @@ impl Connection {
+            };
         }
 
         // Process finished streams list once again, to make sure `Finished`
```
I _think_ this should go before the second DATAGRAM block, as finished events are prioritized over dgrams in the previous block, but it's probably not super important..

LPardue replied:
done

ghedo commented on src/h3/mod.rs:
```diff
@@ -1125,42 +1153,73 @@ impl Connection {
         }
 
-        // Process DATAGRAMs
+        if self.dgram_event_count == self.dgram_poll_threshold &&
```
```suggestion
        // Both stream and DATAGRAM event limits have been reached,
        // so reset and start again.
        if self.dgram_event_count == self.dgram_poll_threshold &&
```

ghedo reviewed (changes requested):
Thanks, the latest logic looks a lot more clear! I added a few comments, but the overall implementation looks good to me.

LPardue commented:
Thanks for the comments. 

I've incorporated them all, and added more description to `Event::Datagram` itself to describe the usage expections in tandem with the new thresholds.

I've also squashed all the commits down.

ghedo commented on src/h3/mod.rs:
```diff
@@ -476,6 +480,46 @@ impl Config {
+    /// an [`Event::Datagram`] and the application should use [`recv_dgram()`]
+    /// to remove it from the pending queue. If all DATAGRAMs are not removed,
+    /// repeat [`poll()`] calls can continue to return events. By
+    /// default there is no threshold to the number times this can happen.
```
```suggestion
    /// default there is no threshold to the number of times this can happen.
```

ghedo commented on src/h3/mod.rs:
```diff
@@ -476,6 +480,46 @@ impl Config {
+    /// an [`Event::Headers`], [`Event::Data`] or [`Event::Finished`]. While
+    /// the stream remains readable
+    /// repeat [`poll()`] calls can continue to return stream events. By
+    /// default there is no threshold to the number times this can happen.
```
```suggestion
    /// default there is no threshold to the number of times this can happen.
```

ghedo reviewed (approved):
LGTM, added a couple of nits, but I will fixe those myself.
